scripts/logid-generator.py

```diff
@@ -85,69 +82,58 @@ def collect_files(directory):
 
   return all_file
 
-def collect_logs(_file, prefix, max_id):
-  ''' Collect logs from the given file, and find the max id within this prefix by the given RE '''
+def collect_all_logs(_file):
+  with open(_file, 'r') as f:
+    logs = re.findall(log_re, f.read())
+    if logs:
+      all_logs.append({'file': _file, "logs": logs})
 
-  global VERSION_SHIFT
-  f = open(_file, 'r')
-
-  for (i, line) in enumerate(f.readlines()):
-    # Check for 0 id
-    if log0_re.search(line) != None:
-      matched_logs.append({'file': _file, 'line_num': i, 'line': line, 'new': ''})
-
-    # Check max id
-    search_res = log_max_re.search(line)
-    if search_res is not None:
-      _id = int(search_res.group('ID'))
-      ids.append({'id': _id, 'file': _file, 'line_num': i})
-      if (_id != 0) and (_id >= prefix and _id < (prefix + VERSION_SHIFT)) and (max_id < _id):
-        max_id = _id
+def max_log_ID(max_id):
+  for logs in all_logs:
+    for log in logs['logs']:
+      if max_id < int(log[2]):
+        max_id = int(log[2])
 
   return max_id
 
 def check_ids(exit_if_zero):
   ''' Find colliding IDs '''
+  count = Counter({'0': 0})
+  for logs in all_logs:
+    count += (Counter(log_ID[2] for log_ID in logs['logs']))
 
-  items = collections.defaultdict(list)
-  for i, item in enumerate(ids):
-    items[item['id']].append(item)
-
-  # Delete items with 0 ID
-  if 0 in items:
-    print("0 ID count:", len(items[0]))
-    del items[0]
-    if exit_if_zero:
-      exit(1)
-
-  # Collect IDs where count > 1
-  duplications = [[_id, data]  for _id, data in items.items()   if len(data) > 1]
-  if len(duplications) > 0:
-    print("ERROR ID duplications", file=sys.stderr)
-    for wrong_id in duplications:
-      print("ID: ", wrong_id[0])
-      for error in wrong_id[1]:
-        print(error)
-    exit(1)
+  duplacated_ids = [k for (k,v) in Counter(count).items() if v > 1]
+  print("0 ID count:", count['0'])
 
-def replace_id(start_id):
-  ''' Replace the 0 id with proper one '''
-
-  _id = start_id
-  for data in matched_logs:
-    #print data['line']
-    data['new'] = log0_re.sub('log_\g<TYPE>(\g<COMPONENT>, %#d, \g<MSG>' % (_id), data['line'])
-    _id += 1
+  if (duplacated_ids or count['0']) and exit_if_zero:
+    for i in duplacated_ids:
+      print("logID:", i, "is duplacated", count[i], "times")
+    exit(1)
 
-def replace_log_in_file(data):
+def replace_log_in_file(file_with_logs, max_id):
   ''' Replace the log in the file '''
+  zero_logIDs = []
+  for log in file_with_logs['logs']:
+    if log[2] == '0':
+      zero_logIDs.append({'message': log[3]})
+
+  if zero_logIDs:
+    prev = ""
+    for i, line in enumerate(fileinput.input(file_with_logs['file'], inplace = 1)):
+      if re.search(r'(\s+|,)0', line):
+        if (
+            any(zero_logID['message'] in line for zero_logID in zero_logIDs)) or \
+            ((re.search('log_', prev) and (any(zero_logID['message'] in prev for zero_logID in zero_logIDs) is False))
```

Comment:
A problem with this is that it can replace 0-s accouring after log lines as well. (But just in files which needs change.)
Take the following example:

>  log_debug(CORE, 1235, "I am debug logging");  // a log line in the previous line
>  int i = 0;   // -> a "0" which will get replaced mistakenly:  int i = 12144545;
>  // ...
>  log_debug(CORE, 0, "I am debug logging again");  // a valid line that needs replacement

Reply:
A proper solution could load the whole file into memory, reuse the "find" regexp for the replacement (if that is possible) and replace in the whole file content.
I do not consider this very critical however in case the logid update is done manually: you can always fix by hand. But if it is not so hard to implement, I think it is worth.